pipeline_dp/dp_engine.py

```diff
@@ -42,3 +42,31 @@ def aggregate(self, col, params: AggregateParams,
     # TODO: implement aggregate().
     # It returns input for now, just to ensure that the an example works.
     return col
+
+  def _bound_cross_partition_contributions(self, col, max_partitions_contributed: int, max_contributions_per_partition: int, aggregator_fn):
+    """
+    Binds the contribution by privacy_id within and across partitions
+    Args:
+      col:collection, with types of each element: (privacy_id, partition_key, value)
+      max_partitions_contributed
+      max_contributions_per_partition
+      aggregator_fn: function that takes a list of values and returns an aggregator object which handles all aggregation logic.
+
+    output:((privacy_id, parition_key), aggregator)
+    """
+    #  per partition-contribution bounding with bounding of each contribution
+    col = self._ops.map_tuple(col, lambda privacy_id, partition_key, v:
+                                        ((privacy_id, partition_key), v),
+                              "To (privacy_id, partition_key), value))")
+    col = self._ops.sample_fixed_per_key(col, max_contributions_per_partition,
+                                      "Sample per (privacy_id, partition_key)")
+
+    # Cross partition bounding
+    col = self._ops.map_tuple(col, lambda pid_pk, v: (pid_pk[0],
+                                                      (pid_pk[1], v)),
+                              "To (privacy_id, (partition_key, vector))")
+    col = self._ops.sample_fixed_per_key(col, max_partitions_contributed,
+                                      "Sample per privacy_id")
+    return self._ops.flat_map(col, lambda pid: [((pid[0], pk_v[0]),
+                                                  aggregator_fn(pk_v[1]))
```

Comment:
Please apply aggregator_fn after 
col = self._ops.sample_fixed_per_key(col, max_contributions_per_partition,

It would simplify code here (no need to have "for-loop" here)

Reply:
I separated out the aggregator_fn call to a separate call.

I think the loop is still needed here. 
`col = self._ops.sample_fixed_per_key(col, max_partitions_contributed,  "Sample per privacy_id")`
returns [pid1, [(pk1, v1), ), (pk2, v2), ...]]
need flat_map and loop in that to convert to
[((pid1, pk1), v1), ((pid1, pk2), v2), ...]

Let me know if I am missing smething.